File: codeforces/contest/max_path_for_each_node.py
from collections import defaultdict
import sys
import threading

def get_int():
    return int(input())
def get_nums():
    return list(map(int, input().split()))
def main():
    n = get_int()
    dp1 = [0]*(n + 1)
    dp2 = [0]*(n + 1) 
    graph = defaultdict(list)
    for _ in range(n-1):
        a, b = get_nums()
        graph[a].append(b)
        graph[b].append(a)
    
    def dfs1(root, parent):
        for child in graph[root]:
            if child != parent:
                dp1[child] = 1 + dp1[root]
                dfs1(child, root)
    def dfs2(root, parent):
        for child in graph[root]:
            if child != parent:
                dp2[child] = 1 + dp2[root]
                dfs2(child, root)
    
                       
    dfs1(1, -1)
    left_end = dp1.index(max(dp1))
    dfs2(left_end, -1)
    right_end = dp2.index(max(dp2))
    dp1 = [0]*(n + 1)
    dfs1(right_end, -1)
    ans = [0]*(n)
    for node in range(1, n + 1):
        ans[node - 1] = max(dp1[node], dp2[node])
    print(*ans)
threading.stack_size(1<<27)
sys.setrecursionlimit(1<<30)
main_thread = threading.Thread(target = main)
main_thread.start()
main_thread.join()

File: codeforces/contest/max_path_for_each_node.cpp
#include <bits/stdc++.h>
 
using namespace std;
 
vector<int> g[200010];
int d1[200010], d2[200010];
 
void dfs1(int i = 0, int p = -1, int d = 0) {
  d1[i] = d;
  for(auto j : g[i])
    if(j != p)
      dfs1(j, i, d + 1);
}
 
void dfs2(int i = 0, int p = -1, int d = 0) {
  d2[i] = d;
  for(auto j : g[i])
    if(j != p)
      dfs2(j, i, d + 1);
}
 
int main() {
  int n;
  cin >> n;
  for(int i = 1; i < n; i++) {
    int a, b;
    cin >> a >> b;
    g[a - 1].push_back(b - 1);
    g[b - 1].push_back(a - 1);
  }
  dfs1();
  dfs2(max_element(d1, d1 + n) - d1);
  dfs1(max_element(d2, d2 + n) - d2);
  for(int i = 0; i < n; i++) {
    cout << max(d1[i], d2[i]) << ' ';
  }
  cout << '\n';
}